### DISCUSSION TITLE
Derive from Collection

### DISCUSSION THREAD

**Discussion by mu88** [xunit/xunit#2261] [Question] [answered]
Hey guys,
I'd like to know why the `Collection` class is marked `sealed`? This prevents us from creating a custom attribute (e. g. `[NeedsDbContext]` instead of `[Collection("NeedsDbContext")]`).

**Comment by bradwilson** [comment, marked as answer]
The answer is: source-based discovery.

The [API that "consumes" `CollectionAttribute`](https://github.com/xunit/xunit/blob/a8614d34999c889e1c8014f679de95d20eae1304/src/xunit.core/Sdk/IXunitTestCollectionFactory.cs#L24) does so via [`ITypeInfo`](https://github.com/xunit/abstractions.xunit/blob/main/src/xunit.abstractions/Reflection/ITypeInfo.cs), which may or may not represent compiled code. It can represent an AST ("Abstract Syntax Tree") over source code inside of a text editor.

As such, you'll find that all of our attributes that are intended to be consumed this way must be sealed, because we must be able to discern their intent simply by inspecting the source code. The way we do so is by looking at the constructor arguments in the source code.

Unfortunately, there's no extensibility mechanism to make what you want simpler at this time.
